components/match2/wikis/clashroyale/match_group_input_custom.lua

```diff
@@ -0,0 +1,704 @@
+			played = true,
+			cards = CustomMatchGroupInput._readCards(map['t' .. opponentIndex .. 'p' .. playerIndex .. 'c']),
+		}
+	end
+end
+
+---@param players table
+---@param opponentIndex integer
+---@param map table
+---@param participants table
+---@return integer
+function CustomMatchGroupInput._processTeamPlayerMapData(players, opponentIndex, map, participants)
+	local tbdIndex = 0
+	local appendIndex = #players + 1
+
+	local playerIndex = 1
+	local playerKey = 't' .. opponentIndex .. 'p' .. playerIndex
+	while playerIndex <= MAX_NUM_PLAYERS_PER_MAP and (String.isNotEmpty(map[playerKey]) or
+		String.isNotEmpty(map[playerKey .. 'link']) or String.isNotEmpty(map[playerKey .. 'c'])) do
+
+		local player = map[playerKey .. 'link'] or map[playerKey]
+		if String.isEmpty(player) or Table.includes({TBD, TBA}, player:lower()) then
+			tbdIndex = tbdIndex + 1
+			player = TBD .. tbdIndex
+		else
+			-- allows fetching the link of the player from preset wiki vars
+			player = mw.ext.TeamLiquidIntegration.resolve_redirect(
+				map[playerKey .. 'link'] or Variables.varDefault(map[playerKey] .. '_page') or map[playerKey]
+			)
+		end
+
+		local playerData = {
+			played = true,
+			cards = CustomMatchGroupInput._readCards(map[playerKey .. 'c']),
+		}
+
+		local match2playerIndex = CustomMatchGroupInput._fetchMatch2PlayerIndexOfPlayer(players, player)
+
+		-- if we have the player not present in match2player add basic data here
+		if not match2playerIndex then
+			match2playerIndex = appendIndex
+			playerData = Table.merge(playerData, {name = player:gsub(' ', '_'), displayname = map[playerKey] or player})
+
+			appendIndex = appendIndex + 1
+		end
+
+		participants[opponentIndex .. '_' .. match2playerIndex] =  playerData
+
+		playerIndex = playerIndex + 1
+		playerKey = 't' .. opponentIndex .. 'p' .. playerIndex
+	end
+
+	return playerIndex - 1
+end
+
+---@param players table
+---@param player string
+---@return integer|nil
+function CustomMatchGroupInput._fetchMatch2PlayerIndexOfPlayer(players, player)
+	local displayNameIndex
+	local displayNameFoundTwice = false
+
+	for match2playerIndex, match2player in pairs(players) do
+		local playerWithUnderscores = player:gsub(' ', '_')
+		if match2player and match2player.name == playerWithUnderscores then
+			return match2playerIndex
+		elseif not displayNameIndex and match2player and match2player.displayname == playerWithUnderscores then
+			displayNameIndex = match2playerIndex
+		elseif match2player and match2player.displayname == playerWithUnderscores then
+			displayNameFoundTwice = true
+		end
+	end
+
+	if not displayNameFoundTwice then
+		return displayNameIndex
+	end
+end
+
+---@param input string
+---@return table
+function CustomMatchGroupInput._readCards(input)
+	local cards = Json.parseIfString(input) or {}
+
+	for cardIndex, card in pairs(cards) do
+		if not CardNames[card:lower()] then
+			error('Invalid Card "' .. card .. '"')
+		end
+		cards[cardIndex] = CardNames[card]
+	end
+
+	return cards
```

Comment:
```suggestion
	return Array.map(Json.parseIfString(input) or {}, function(card)
		if String.isEmpty(card) then return EMPTY_CARD end
		local cleanedCard =  CardNames[card]
		assert(cleanedCard, 'Invalid Card "' .. card .. '"')
		return cleanedCard
	)
```

Reply:
Since ``EMPTY_CARD`` is undefined (per the linter), what should we be defining this as to avoid causing issues?

Reply:
if in the card display module there is a placeholder you could use that

or just use nil instead

Reply:
Looks like '' is the correct output